### PR TITLE
Introduce a new `RoomProxyType` and treat rooms differently based on their membership state

### PR DESCRIPTION
**Pull request by stefanceriu**

### CODE REVIEW

File: ElementX/Sources/Screens/HomeScreen/HomeScreenViewModel.swift [hunk 1]
```diff
@@ -415,19 +413,16 @@ class HomeScreenViewModel: HomeScreenViewModelType, HomeScreenViewModelProtocol
             }
             userIndicatorController.submitIndicator(UserIndicator(id: Self.leaveRoomLoadingID, type: .modal, title: L10n.commonLeavingRoom, persistent: true))
             
-            let room = await userSession.clientProxy.roomForIdentifier(roomId)
-            let result = await room?.leaveRoom()
```
Comment (pixlwave) [inline ElementX/Sources/Screens/HomeScreen/HomeScreenViewModel.swift:419]: Err, don't you need to call this on the joined room?

Also unrelated but *cough* `roomID` *cough*

File: ElementX/Sources/Screens/JoinRoomScreen/JoinRoomScreenViewModel.swift [hunk 2]
```diff
@@ -27,7 +27,7 @@ class JoinRoomScreenViewModel: JoinRoomScreenViewModelType, JoinRoomScreenViewMo
     private let userIndicatorController: UserIndicatorControllerProtocol
     
     private var roomPreviewDetails: RoomPreviewDetails?
-    private var roomProxy: RoomProxyProtocol?
+    private var room: RoomProxyType?
```
Comment (pixlwave) [inline ElementX/Sources/Screens/JoinRoomScreen/JoinRoomScreenViewModel.swift:30]: Maybe I'm missing something here, but why would this be anything other than an `InvitedRoomProxyProtocol`? Then we wouldn't need to fetch it twice in this screen.

File: ElementX/Sources/Services/Room/JoinedRoomProxy.swift [hunk 3]
```diff
@@ -77,13 +77,17 @@ class RoomProxy: RoomProxyProtocol {
...
     }
         
-    private let actionsSubject = PassthroughSubject<RoomProxyAction, Never>()
-    var actionsPublisher: AnyPublisher<RoomProxyAction, Never> {
+    private let actionsSubject = PassthroughSubject<JoinedRoomProxyAction, Never>()
+    var actionsPublisher: AnyPublisher<JoinedRoomProxyAction, Never> {
         actionsSubject.eraseToAnyPublisher()
     }
     
     lazy var id: String = room.id()
```
Comment (pixlwave) [inline ElementX/Sources/Services/Room/JoinedRoomProxy.swift:85]: Is it a dumb question to ask why this one is still lazy?

File: ElementX/Sources/Services/Room/InvitedRoomProxy.swift [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+
+import Foundation
+import MatrixRustSDK
+import UIKit
+
+class InvitedRoomProxy: InvitedRoomProxyProtocol {
+    private let roomListItem: RoomListItemProtocol
+    private let room: RoomProtocol
+    
+    lazy var id: String = room.id()
```
Comment (pixlwave) [inline ElementX/Sources/Services/Room/InvitedRoomProxy.swift:25]: I guess same question here.

File: ElementX/Sources/Services/Client/ClientProxy.swift [hunk 5]
```diff
@@ -868,19 +844,46 @@ class ClientProxy: ClientProxyProtocol {
...
+        }
+        
+        guard let roomListItem = try? roomListService.room(roomId: identifier) else {
+            MXLog.error("Failed retrieving room, invalid room list item")
+            return nil
+        }
+        
+        if roomListItem.isTimelineInitialized() == false {
+            do {
+                try await roomListItem.initTimeline(eventTypeFilter: eventFilters, internalIdPrefix: nil)
```
Comment (pixlwave) [inline ElementX/Sources/Services/Client/ClientProxy.swift:861]: Wait so does this now work for a room invite too?

**Review by pixlwave — APPROVED** [review]
Very nice tidy-up, this is much safer 👏👏👏

Did it need any Rust changes in the end?

File: ElementX/Sources/Services/Client/ClientProxy.swift (hunk 5 above)
Comment (stefanceriu) [reply ElementX/Sources/Services/Client/ClientProxy.swift:861]: Nop, that's still completely broken 😁

File: ElementX/Sources/Services/Room/JoinedRoomProxy.swift (hunk 3 above)
Comment (stefanceriu) [reply ElementX/Sources/Services/Room/JoinedRoomProxy.swift:85]: No idea to be honest ¯\_(ツ)_/¯

File: ElementX/Sources/Screens/JoinRoomScreen/JoinRoomScreenViewModel.swift (hunk 2 above)
Comment (stefanceriu) [reply ElementX/Sources/Screens/JoinRoomScreen/JoinRoomScreenViewModel.swift:30]: Talked about it offline, the sort answer is: to pave the way for knocked rooms.

**Pull request merged**